http_parser.c

```diff
@@ -2367,8 +2367,22 @@ http_parser_parse_url(const char *buf, size_t buflen, int is_connect,
   }
 
   if (u->field_set & (1 << UF_PORT)) {
-    /* Don't bother with endp; we've already validated the string */
-    unsigned long v = strtoul(buf + u->field_data[UF_PORT].off, NULL, 10);
+    uint16_t off;
+    uint16_t len;
+    const char* end;
+    unsigned long v;
+
+    off = u->field_data[UF_PORT].off;
+    len = u->field_data[UF_PORT].len;
+    end = buf + off + len;
+
+    /* NOTE: The characters are already validated and are in the [0-9] range */
+    assert(off + len <= buflen && "Port number overflow");
+    v = 0;
+    for (const char* p = buf + off; p < end; p++) {
+      v *= 10;
+      v += *p - '0';
+    }
 
     /* Ports have a max value of 2^16 */
     if (v > 0xffff) {
```

Comment:
You could move this check inside the loop to protect against `v` wrapping around.

Just a suggestion though; with this PR it's already better than before because errno no longer gets clobbered.

Reply:
Ack.